doc/lua_api.txt

```diff
@@ -4557,6 +4557,23 @@ Utilities
 * `minetest.colorspec_to_colorstring(colorspec)`: Converts a ColorSpec to a
   ColorString. If the ColorSpec is invalid, returns `nil`.
     * `colorspec`: The ColorSpec to convert
+* `minetest.colorspec_to_bytes(colorspec)`: Converts a ColorSpec to a raw
+  string of four bytes in an RGBA layout, returned as a string.
+  * `colorspec`: The ColorSpec to convert
+* `minetest.encode_png(width, height, data, [compression])`: Encode a PNG
+  image and return it in string form.
+    * `width`: Width of the image
+    * `height`: Height of the image
+    * `data`: Image data, one of:
+        * array table of ColorSpec, length must be width*height
+        * string with raw RGBA pixels, length must be width*height*4
+    * `compression`: Optional zlib compression level, number in range 0 to 9.
+  The data is one-dimensional, starting in the upper left corner of the image
+  and laid out in scanlines going from left to right, then top to bottom.
+  Please note that it's not safe to use string.char to generate raw data,
```

Comment:
Huh, why is that?

Reply:
The `string.char` thing? I have no idea really, it just produced a broken result while `colorspec_to_bytes` did not. I think it was having trouble with the 0xFF byte. The Lua manual states that string.char output is not portable between platforms so I decided to just condemn it as unsafe.

`ffi.string` is probably the best method to produce the buffer but not everyone can use it.

Reply:
btw this worked for me:
![grafik](https://user-images.githubusercontent.com/1042418/127000549-70ad1dae-9655-42d6-9388-3e13192b8f7c.png)

Reply:
At one point I had something in the test cases like:
```
data_cs[i] = {
	r = floor(fractal[i] * 255),
	g = floor(abs(fractal[i] * 2 - 1) * 255),
	b = floor(abs(1 - fractal[i]) * 255),
	a = 255,
}
data_raw[i] = string.char( data_cs[i].r, data_cs[i].g, data_cs[i].b, data_cs[i].a )
```
where data_raw was later fed to `table.concat` like you did there for one image, and data_cs was used for another. In theory the output should have been identical for both images, because the colorspec converter in principle does exactly the same thing as `string.char`.

Instead of being identical, the raw version of the image had 50% transparency in all but a few pixels, and a few odd pixels in some locations, suggesting that the 0xFF byte was being mangled to 0x80 or 0x7F, somehow.

This may be locale and lua runtime dependent so I think it's best if string.char is avoided.

Reply:
Sure, it's fine if the advice in the docs is kept.